extensions/amp-story/1.0/amp-story-embedded-component.js

```diff
@@ -534,30 +559,93 @@ export class AmpStoryEmbeddedComponent {
         /** measure */
         () => {
           const targetRect = target./*OK*/getBoundingClientRect();
+          // TODO(#20832): Store DOMRect for the page in the store to avoid
+          // having to call getBoundingClientRect().
           const pageRect = this.componentPage_./*OK*/getBoundingClientRect();
 
-          const centeredTop = pageRect.height / 2 - targetRect.height / 2;
-          const centeredLeft = pageRect.width / 2 - targetRect.width / 2;
-
-          // Only account for offset from target to page borders. Since in
-          // desktop mode page is not at the borders of viewport.
-          const leftOffset = targetRect.left - pageRect.left;
-          const topOffset = targetRect.top - pageRect.top;
-
-          state.translateY = centeredTop - topOffset;
-          state.translateX = leftOffset - centeredLeft;
+          // Retrieve scale() previously set by prepareForAnimation() on the
+          // element.
+          this.scaleFactor_ =
+            parseFloat(getStyle(target, 'transform').split('scale(')[1]);
+          const fullScreenWidth = targetRect.width / this.scaleFactor_;
+          // Gap on the left of the element between full-screen size and
+          // current size.
+          const leftGap = (fullScreenWidth - targetRect.width) / 2;
+          // Distance from left of page to what will be the left of the
+          // element in full-screen.
+          const fullScreenLeft = targetRect.left - leftGap - pageRect.left;
+          const centeredLeft = pageRect.width / 2 - fullScreenWidth / 2;
+          state.translateX = centeredLeft - fullScreenLeft;
+
+          // Calculate what the height of the element will be in full-screen.
+          const fullScreenHeight = targetRect.height / this.scaleFactor_;
+          // Gap on the top of the element between full-screen size and
+          // current size.
+          const topGap = (fullScreenHeight - targetRect.height) / 2;
+          // Distance from top of page to what will be the top of the element in
+          // full-screen.
+          const fullScreenTop = targetRect.top - topGap - pageRect.top;
+          const centeredTop = pageRect.height / 2 - fullScreenHeight / 2;
+          state.translateY = centeredTop - fullScreenTop;
         },
         /** mutate */
         () => {
           target.classList.add('i-amphtml-animate-expand-in');
+          target.classList.toggle('i-amphtml-expanded-component', true);
```

Comment:
Re: CSS classes, would be possible to, eventually, factorize `i-amphtml-expanded-mode`, `i-amphtml-expanded-component`, and `i-amphtml-animate-expand-in` as a single class?